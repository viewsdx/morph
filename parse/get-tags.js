import { isAnimation, isStyle, isUnsupportedShorthand } from './helpers.js'

const SLOT_PROPS = ['from', 'when', 'onClick', 'onFocus', 'onWhen']
const shouldBeSlot = prop => SLOT_PROPS.includes(prop) || /^on[A-Z]/.test(prop)

export default ({ name, isSlot, slotName, slotIsNot, value, block }) => {
  const tags = {}

  if (isAnimation(value) && name !== 'text') tags.animation = true
  if (isStyle(name)) tags.style = true
  if (isUnsupportedShorthand(name) && block.isBasic) {
    tags.unsupportedShorthand = true
  }

  if (shouldBeSlot(name)) tags.shouldBeSlot = true
  if (isSlot) tags.slot = true
  if (slotIsNot) tags.slotIsNot = true

  tags.validSlot = tags.slot || (tags.shouldBeSlot && tags.slot) || null

  return tags
}
