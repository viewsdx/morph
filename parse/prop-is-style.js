export const STYLE = [
  'alignContent',
  'alignItems',
  'alignSelf',
  'appRegion',
  'background',
  'backgroundColor',
  'backgroundImage',
  'backgroundSize',
  'borderBottomColor',
  'borderBottomLeftRadius',
  'borderBottomRightRadius',
  'borderBottomStyle',
  'borderBottomWidth',
  'borderColor',
  'borderLeftColor',
  'borderLeftStyle',
  'borderLeftWidth',
  'borderRightColor',
  'borderRightStyle',
  'borderRightWidth',
  'borderStyle',
  'borderTopColor',
  'borderTopLeftRadius',
  'borderTopRightRadius',
  'borderTopStyle',
  'borderTopWidth',
  'borderWidth',
  'bottom',
  'clipPath',
  'color',
  'cursor',
  'display',
  'fill',
  'filter',
  'flexGrow',
  'flexBasis',
  'flexDirection',
  'flexFlow',
  'flexShrink',
  'flexWrap',
  'fontFamily',
  'fontSize',
  'fontStyle',
  'fontWeight',
  'height',
  'hyphens',
  'justifyContent',
  'left',
  'letterSpacing',
  'lineHeight',
  'marginBottom',
  'marginLeft',
  'marginRight',
  'marginTop',
  'maxHeight',
  'maxWidth',
  'minHeight',
  'minWidth',
  'opacity',
  'outlineColor',
  'outlineOffset',
  'outlineStyle',
  'outlineWidth',
  'overflowWrap',
  'overflowX',
  'overflowY',
  'paddingBottom',
  'paddingLeft',
  'paddingRight',
  'paddingTop',
  'pointerEvents',
  'position',
  'resize',
  'right',
  'shadowColor',
  'shadowOffsetX',
  'shadowOffsetY',
  'shadowBlur',
  'shadowSpread',
  'stroke',
  'textAlign',
  'textDecoration',
  'textOverflow',
  'textJustify',
  'textTransform',
  'top',
  'translateX',
  'translateY',
  'scale',
  'rotate',
  'rotateX',
  'rotateY',
  'transformOriginX',
  'transformOriginY',
  'transition',
  'userSelect',
  'whiteSpace',
  'width',
  'wordBreak',
  'wordSpacing',
  'wordWrap',
  'zIndex',
  // '@fontFace',
  // '@keyframes',
  // '@media',
  // 'background',
  // 'backgroundAttachment',
  // 'backgroundBlend-mode',
  // 'backgroundClip',
  // 'backgroundOrigin',
  // 'backgroundPosition',
  // 'backgroundRepeat',
  // 'border',
  // 'borderBottom',
  // 'borderCollapse',
  // 'borderImage',
  // 'borderImageOutset',
  // 'borderImageRepeat',
  // 'borderImageSlice',
  // 'borderImageSource',
  // 'borderImageWidth',
  // 'borderLeft',
  // 'borderRadius',
  // 'borderRight',
  // 'borderSpacing',
  // 'borderTop',
  // 'boxShadow',
  // 'boxSizing',
  // 'captionSide',
  // 'clear',
  // 'clip',
  // 'column-gap',
  // 'columnCount',
  // 'columnFill',
  // 'columnRule',
  // 'columnRuleColor',
  // 'columnRuleStyle',
  // 'columnRuleWidth',
  // 'columnSpan',
  // 'columnWidth',
  // 'columns',
  // 'content',
  // 'counterIncrement',
  // 'counterReset',
  // 'direction',
  // 'display',
  // 'emptyCells',
  // 'float',
  // 'font',
  // 'font-variant',
  // 'fontSizeAdjust',
  // 'fontStretch',
  // 'hangingPunctuation',
  // 'listStyle',
  // 'listStyleImage',
  // 'listStylePosition',
  // 'listStyleType',
  // 'margin',
  // 'nav-up',
  // 'navDown',
  // 'navIndex',
  // 'navLeft',
  // 'navRight',
  // 'order',
  // 'outline',
  // 'overflow',
  // 'padding',
  // 'pageBreakAfter',
  // 'pageBreakBefore',
  // 'pageBreakInside',
  // 'perspective',
  // 'perspectiveOrigin',
  // 'quotes',
  // 'tabSize',
  // 'tableLayout',
  // 'textAlignLast',
  // 'textDecorationColor',
  // 'textDecorationLine',
  // 'textDecorationStyle',
  // 'textIndent',
  // 'textShadow',
  // 'transformStyle',
  // 'transitionDelay',
  // 'transitionDuration',
  // 'transitionProperty',
  // 'transitionTimingFunction',
  // 'unicodeBidi',
  // 'verticalAlign',
  // 'visibility',
]

export const isStyle = name => STYLE.includes(name)
