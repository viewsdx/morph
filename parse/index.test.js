import parse from './index.js'

test('#parse', () => {
  VIEWS.forEach(view => expect(parse({ source: view })).toMatchSnapshot())

  expect(parse({ source: WARNING })).toMatchSnapshot()
})

const VIEWS = [
  `Main Vertical`,

  `Main Vertical
Text`,

  `BlueButton Vertical
Before Text

Nested Vertical
ImageInside Image
source https://image.com/file.jpg

VerticalInside Vertical


After Text
text after
color purple
when hover
color red

Last Vertical
backgroundColor blue`,

  // ensure that empty spaces at the end of the block don't block the block from
  // being recognised as a stop point for the props of the previous block
  `A Vertical 
A1 Text
color red

A2 Text    
color white`,

  `Vertical
backgroundColor <backgroundColor
onClick <onClick
Title Text
color < red
marginLeft < 10
text <title This is the title

CaptureText
onFocus <

List
from <
Stuff`,
]

const WARNING = `Warning Vertical
EmptyWhen Vertical
color blue
marginTop <
border 1px solid red
when
color red
when < stuff
backgroundColor purple
color green`
