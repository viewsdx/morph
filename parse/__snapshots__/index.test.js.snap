// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#parse 1`] = `
Object {
  "fonts": Array [],
  "locals": Array [],
  "slots": Array [],
  "views": Array [
    Object {
      "children": Array [],
      "is": "Main",
      "isBasic": true,
      "isGroup": true,
      "isList": false,
      "loc": Object {
        "end": Object {
          "column": 12,
          "line": 0,
        },
        "start": Object {
          "column": 0,
          "line": 0,
        },
      },
      "name": "Vertical",
      "properties": Array [],
      "scopes": Array [],
      "type": "Block",
    },
  ],
  "warnings": Array [],
}
`;

exports[`#parse 2`] = `
Object {
  "fonts": Array [],
  "locals": Array [],
  "slots": Array [],
  "views": Array [
    Object {
      "children": Array [
        Object {
          "isBasic": true,
          "isGroup": false,
          "loc": Object {
            "end": Object {
              "column": 3,
              "line": 1,
            },
            "start": Object {
              "column": 0,
              "line": 1,
            },
          },
          "name": "Text",
          "properties": Array [],
          "scopes": Array [],
          "type": "Block",
        },
      ],
      "is": "Main",
      "isBasic": true,
      "isGroup": true,
      "isList": false,
      "loc": Object {
        "end": Object {
          "column": 3,
          "line": 1,
        },
        "start": Object {
          "column": 0,
          "line": 0,
        },
      },
      "name": "Vertical",
      "properties": Array [],
      "scopes": Array [],
      "type": "Block",
    },
  ],
  "warnings": Array [],
}
`;

exports[`#parse 3`] = `
Object {
  "fonts": Array [],
  "locals": Array [],
  "slots": Array [],
  "views": Array [
    Object {
      "children": Array [
        Object {
          "is": "Before",
          "isBasic": true,
          "isGroup": false,
          "loc": Object {
            "end": Object {
              "column": 0,
              "line": 2,
            },
            "start": Object {
              "column": 0,
              "line": 1,
            },
          },
          "name": "Text",
          "properties": Array [],
          "scopes": Array [],
          "type": "Block",
        },
        Object {
          "children": Array [
            Object {
              "is": "ImageInside",
              "isBasic": true,
              "isGroup": false,
              "loc": Object {
                "end": Object {
                  "column": 0,
                  "line": 6,
                },
                "start": Object {
                  "column": 0,
                  "line": 4,
                },
              },
              "name": "Image",
              "properties": Array [
                Object {
                  "loc": Object {
                    "end": Object {
                      "column": 32,
                      "line": 5,
                    },
                    "start": Object {
                      "column": 0,
                      "line": 5,
                    },
                  },
                  "name": "source",
                  "tags": Object {
                    "validSlot": null,
                  },
                  "type": "Property",
                  "value": "https://image.com/file.jpg",
                },
              ],
              "scopes": Array [],
              "type": "Block",
            },
            Object {
              "children": Array [],
              "is": "VerticalInside",
              "isBasic": true,
              "isGroup": true,
              "isList": false,
              "loc": Object {
                "end": Object {
                  "column": 0,
                  "line": 8,
                },
                "start": Object {
                  "column": 0,
                  "line": 7,
                },
              },
              "name": "Vertical",
              "properties": Array [],
              "scopes": Array [],
              "type": "Block",
            },
          ],
          "is": "Nested",
          "isBasic": true,
          "isGroup": true,
          "isList": false,
          "loc": Object {
            "end": Object {
              "column": 0,
              "line": 9,
            },
            "start": Object {
              "column": 0,
              "line": 3,
            },
          },
          "name": "Vertical",
          "properties": Array [],
          "scopes": Array [],
          "type": "Block",
        },
        Object {
          "is": "After",
          "isBasic": true,
          "isGroup": false,
          "loc": Object {
            "end": Object {
              "column": 0,
              "line": 15,
            },
            "start": Object {
              "column": 0,
              "line": 10,
            },
          },
          "name": "Text",
          "properties": Array [
            Object {
              "loc": Object {
                "end": Object {
                  "column": 9,
                  "line": 11,
                },
                "start": Object {
                  "column": 0,
                  "line": 11,
                },
              },
              "name": "text",
              "tags": Object {
                "validSlot": null,
              },
              "type": "Property",
              "value": "after",
            },
            Object {
              "loc": Object {
                "end": Object {
                  "column": 11,
                  "line": 12,
                },
                "start": Object {
                  "column": 0,
                  "line": 12,
                },
              },
              "name": "color",
              "tags": Object {
                "style": true,
                "validSlot": null,
              },
              "type": "Property",
              "value": "purple",
            },
          ],
          "scopes": Array [
            Object {
              "isLocal": false,
              "isSystem": true,
              "name": "when",
              "properties": Array [
                Object {
                  "loc": Object {
                    "end": Object {
                      "column": 10,
                      "line": 13,
                    },
                    "start": Object {
                      "column": 0,
                      "line": 13,
                    },
                  },
                  "name": "when",
                  "slotName": "hover",
                  "tags": Object {
                    "scope": "<hover",
                    "shouldBeSlot": true,
                    "slot": true,
                    "validSlot": true,
                  },
                  "type": "Property",
                  "value": "hover",
                },
                Object {
                  "loc": Object {
                    "end": Object {
                      "column": 8,
                      "line": 14,
                    },
                    "start": Object {
                      "column": 0,
                      "line": 14,
                    },
                  },
                  "name": "color",
                  "tags": Object {
                    "style": true,
                    "validSlot": null,
                  },
                  "type": "Property",
                  "value": "red",
                },
              ],
              "slotIsNot": false,
              "slotName": "hover",
              "value": "hover",
            },
          ],
          "type": "Block",
        },
        Object {
          "children": Array [],
          "is": "Last",
          "isBasic": true,
          "isGroup": true,
          "isList": false,
          "loc": Object {
            "end": Object {
              "column": 19,
              "line": 17,
            },
            "start": Object {
              "column": 0,
              "line": 16,
            },
          },
          "name": "Vertical",
          "properties": Array [
            Object {
              "loc": Object {
                "end": Object {
                  "column": 19,
                  "line": 17,
                },
                "start": Object {
                  "column": 0,
                  "line": 17,
                },
              },
              "name": "backgroundColor",
              "tags": Object {
                "style": true,
                "validSlot": null,
              },
              "type": "Property",
              "value": "blue",
            },
          ],
          "scopes": Array [],
          "type": "Block",
        },
      ],
      "is": "BlueButton",
      "isBasic": true,
      "isGroup": true,
      "isList": false,
      "loc": Object {
        "end": Object {
          "column": 19,
          "line": 17,
        },
        "start": Object {
          "column": 0,
          "line": 0,
        },
      },
      "name": "Vertical",
      "properties": Array [],
      "scopes": Array [],
      "type": "Block",
    },
  ],
  "warnings": Array [],
}
`;

exports[`#parse 4`] = `
Object {
  "fonts": Array [],
  "locals": Array [],
  "slots": Array [],
  "views": Array [
    Object {
      "children": Array [
        Object {
          "is": "A1",
          "isBasic": true,
          "isGroup": false,
          "loc": Object {
            "end": Object {
              "column": 0,
              "line": 3,
            },
            "start": Object {
              "column": 0,
              "line": 1,
            },
          },
          "name": "Text",
          "properties": Array [
            Object {
              "loc": Object {
                "end": Object {
                  "column": 8,
                  "line": 2,
                },
                "start": Object {
                  "column": 0,
                  "line": 2,
                },
              },
              "name": "color",
              "tags": Object {
                "style": true,
                "validSlot": null,
              },
              "type": "Property",
              "value": "red",
            },
          ],
          "scopes": Array [],
          "type": "Block",
        },
        Object {
          "is": "A2",
          "isBasic": true,
          "isGroup": false,
          "loc": Object {
            "end": Object {
              "column": 10,
              "line": 5,
            },
            "start": Object {
              "column": 0,
              "line": 4,
            },
          },
          "name": "Text",
          "properties": Array [
            Object {
              "loc": Object {
                "end": Object {
                  "column": 10,
                  "line": 5,
                },
                "start": Object {
                  "column": 0,
                  "line": 5,
                },
              },
              "name": "color",
              "tags": Object {
                "style": true,
                "validSlot": null,
              },
              "type": "Property",
              "value": "white",
            },
          ],
          "scopes": Array [],
          "type": "Block",
        },
      ],
      "is": "A",
      "isBasic": true,
      "isGroup": true,
      "isList": false,
      "loc": Object {
        "end": Object {
          "column": 10,
          "line": 5,
        },
        "start": Object {
          "column": 0,
          "line": 0,
        },
      },
      "name": "Vertical",
      "properties": Array [],
      "scopes": Array [],
      "type": "Block",
    },
  ],
  "warnings": Array [
    Object {
      "line": "A Vertical ",
      "loc": Object {
        "start": Object {
          "line": 0,
        },
      },
      "type": "You have some spaces before or after this line. Clean them up.",
    },
    Object {
      "line": "A2 Text    ",
      "loc": Object {
        "start": Object {
          "line": 4,
        },
      },
      "type": "You have some spaces before or after this line. Clean them up.",
    },
  ],
}
`;

exports[`#parse 5`] = `
Object {
  "fonts": Array [],
  "locals": Array [],
  "slots": Array [
    Object {
      "defaultValue": false,
      "name": "backgroundColor",
      "type": "string",
    },
    Object {
      "defaultValue": false,
      "name": "onClick",
      "type": "function",
    },
    Object {
      "defaultValue": "red",
      "name": "color",
      "type": "string",
    },
    Object {
      "defaultValue": 10,
      "name": "marginLeft",
      "type": "number",
    },
    Object {
      "defaultValue": "This is the title",
      "name": "title",
      "type": "string",
    },
    Object {
      "defaultValue": false,
      "name": "onFocus",
      "type": "function",
    },
    Object {
      "defaultValue": false,
      "name": "from",
      "type": "array",
    },
  ],
  "views": Array [
    Object {
      "children": Array [
        Object {
          "is": "Title",
          "isBasic": true,
          "isGroup": false,
          "loc": Object {
            "end": Object {
              "column": 0,
              "line": 7,
            },
            "start": Object {
              "column": 0,
              "line": 3,
            },
          },
          "maybeAnimated": true,
          "name": "Text",
          "properties": Array [
            Object {
              "defaultValue": "red",
              "loc": Object {
                "end": Object {
                  "column": 10,
                  "line": 4,
                },
                "start": Object {
                  "column": 0,
                  "line": 4,
                },
              },
              "name": "color",
              "slotName": "",
              "tags": Object {
                "slot": true,
                "style": true,
                "validSlot": true,
              },
              "type": "Property",
              "value": "props.color",
            },
            Object {
              "defaultValue": 10,
              "loc": Object {
                "end": Object {
                  "column": 14,
                  "line": 5,
                },
                "start": Object {
                  "column": 0,
                  "line": 5,
                },
              },
              "name": "marginLeft",
              "slotName": "",
              "tags": Object {
                "slot": true,
                "style": true,
                "validSlot": true,
              },
              "type": "Property",
              "value": "props.marginLeft",
            },
            Object {
              "defaultValue": "This is the title",
              "loc": Object {
                "end": Object {
                  "column": 28,
                  "line": 6,
                },
                "start": Object {
                  "column": 0,
                  "line": 6,
                },
              },
              "name": "text",
              "slotName": "title",
              "tags": Object {
                "slot": true,
                "validSlot": true,
              },
              "type": "Property",
              "value": "props.title",
            },
          ],
          "scopes": Array [],
          "type": "Block",
        },
        Object {
          "isBasic": true,
          "isGroup": false,
          "loc": Object {
            "end": Object {
              "column": 0,
              "line": 10,
            },
            "start": Object {
              "column": 0,
              "line": 8,
            },
          },
          "name": "CaptureText",
          "properties": Array [
            Object {
              "defaultValue": "",
              "loc": Object {
                "end": Object {
                  "column": 8,
                  "line": 9,
                },
                "start": Object {
                  "column": 0,
                  "line": 9,
                },
              },
              "name": "onFocus",
              "slotName": "",
              "tags": Object {
                "shouldBeSlot": true,
                "slot": true,
                "validSlot": true,
              },
              "type": "Property",
              "value": "props.onFocus",
            },
          ],
          "scopes": Array [],
          "type": "Block",
        },
        Object {
          "children": Array [
            Object {
              "children": Array [],
              "childrenProxyMap": null,
              "isBasic": false,
              "isGroup": true,
              "isList": false,
              "loc": Object {
                "end": Object {
                  "column": 4,
                  "line": 13,
                },
                "start": Object {
                  "column": 0,
                  "line": 13,
                },
              },
              "name": "Stuff",
              "properties": Array [],
              "scopes": Array [],
              "type": "Block",
            },
          ],
          "isBasic": true,
          "isGroup": true,
          "isList": true,
          "loc": Object {
            "end": Object {
              "column": 4,
              "line": 13,
            },
            "start": Object {
              "column": 0,
              "line": 11,
            },
          },
          "name": "List",
          "properties": Array [
            Object {
              "defaultValue": "",
              "loc": Object {
                "end": Object {
                  "column": 5,
                  "line": 12,
                },
                "start": Object {
                  "column": 0,
                  "line": 12,
                },
              },
              "name": "from",
              "slotName": "",
              "tags": Object {
                "shouldBeSlot": true,
                "slot": true,
                "validSlot": true,
              },
              "type": "Property",
              "value": "props.from",
            },
          ],
          "scopes": Array [],
          "type": "Block",
        },
      ],
      "isBasic": true,
      "isGroup": true,
      "isList": false,
      "loc": Object {
        "end": Object {
          "column": 4,
          "line": 13,
        },
        "start": Object {
          "column": 0,
          "line": 0,
        },
      },
      "maybeAnimated": true,
      "name": "Vertical",
      "properties": Array [
        Object {
          "defaultValue": false,
          "loc": Object {
            "end": Object {
              "column": 31,
              "line": 1,
            },
            "start": Object {
              "column": 0,
              "line": 1,
            },
          },
          "name": "backgroundColor",
          "slotName": "backgroundColor",
          "tags": Object {
            "slot": true,
            "style": true,
            "validSlot": true,
          },
          "type": "Property",
          "value": "props.backgroundColor",
        },
        Object {
          "defaultValue": "onClick",
          "loc": Object {
            "end": Object {
              "column": 15,
              "line": 2,
            },
            "start": Object {
              "column": 0,
              "line": 2,
            },
          },
          "name": "onClick",
          "slotName": "onClick",
          "tags": Object {
            "shouldBeSlot": true,
            "slot": true,
            "validSlot": true,
          },
          "type": "Property",
          "value": "props.onClick",
        },
      ],
      "scopes": Array [],
      "type": "Block",
    },
  ],
  "warnings": Array [
    Object {
      "line": "backgroundColor <backgroundColor",
      "loc": Object {
        "end": Object {
          "column": 31,
          "line": 1,
        },
        "start": Object {
          "column": 0,
          "line": 1,
        },
      },
      "type": "Add a default value to \\"backgroundColor\\" like: \\"backgroundColor <backgroundColor default value\\"",
    },
  ],
}
`;

exports[`#parse 6`] = `
Object {
  "fonts": Array [],
  "locals": Array [
    "es",
  ],
  "slots": Array [],
  "views": Array [
    Object {
      "children": Array [
        Object {
          "isBasic": true,
          "isGroup": false,
          "loc": Object {
            "end": Object {
              "column": 8,
              "line": 4,
            },
            "start": Object {
              "column": 0,
              "line": 1,
            },
          },
          "name": "Text",
          "properties": Array [
            Object {
              "loc": Object {
                "end": Object {
                  "column": 6,
                  "line": 2,
                },
                "start": Object {
                  "column": 0,
                  "line": 2,
                },
              },
              "name": "text",
              "tags": Object {
                "validSlot": null,
              },
              "type": "Property",
              "value": "hi",
            },
          ],
          "scopes": Array [
            Object {
              "isLocal": true,
              "isSystem": false,
              "name": "when",
              "properties": Array [
                Object {
                  "defaultValue": "es",
                  "loc": Object {
                    "end": Object {
                      "column": 7,
                      "line": 3,
                    },
                    "start": Object {
                      "column": 0,
                      "line": 3,
                    },
                  },
                  "name": "when",
                  "slotName": "es",
                  "tags": Object {
                    "scope": "<es",
                    "shouldBeSlot": true,
                    "slot": true,
                    "validSlot": true,
                  },
                  "type": "Property",
                  "value": "props.es",
                },
                Object {
                  "loc": Object {
                    "end": Object {
                      "column": 8,
                      "line": 4,
                    },
                    "start": Object {
                      "column": 0,
                      "line": 4,
                    },
                  },
                  "name": "text",
                  "tags": Object {
                    "validSlot": null,
                  },
                  "type": "Property",
                  "value": "hola",
                },
              ],
              "slotIsNot": false,
              "slotName": "es",
              "value": "es",
            },
          ],
          "type": "Block",
        },
      ],
      "is": "Locals",
      "isBasic": true,
      "isGroup": true,
      "isList": false,
      "loc": Object {
        "end": Object {
          "column": 8,
          "line": 4,
        },
        "start": Object {
          "column": 0,
          "line": 0,
        },
      },
      "name": "Vertical",
      "properties": Array [],
      "scopes": Array [],
      "type": "Block",
    },
  ],
  "warnings": Array [],
}
`;

exports[`#parse 7`] = `
Object {
  "fonts": Array [],
  "locals": Array [],
  "slots": Array [
    Object {
      "defaultValue": false,
      "name": "marginTop",
      "type": "number",
    },
  ],
  "views": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [],
          "is": "EmptyWhen",
          "isBasic": true,
          "isGroup": true,
          "isList": false,
          "loc": Object {
            "end": Object {
              "column": 10,
              "line": 9,
            },
            "start": Object {
              "column": 0,
              "line": 1,
            },
          },
          "maybeAnimated": true,
          "name": "Vertical",
          "properties": Array [
            Object {
              "loc": Object {
                "end": Object {
                  "column": 9,
                  "line": 2,
                },
                "start": Object {
                  "column": 0,
                  "line": 2,
                },
              },
              "name": "color",
              "tags": Object {
                "style": true,
                "validSlot": null,
              },
              "type": "Property",
              "value": "blue",
            },
            Object {
              "defaultValue": false,
              "loc": Object {
                "end": Object {
                  "column": 10,
                  "line": 3,
                },
                "start": Object {
                  "column": 0,
                  "line": 3,
                },
              },
              "name": "marginTop",
              "slotName": "",
              "tags": Object {
                "slot": true,
                "style": true,
                "validSlot": true,
              },
              "type": "Property",
              "value": "props.marginTop",
            },
            Object {
              "loc": Object {
                "end": Object {
                  "column": 19,
                  "line": 4,
                },
                "start": Object {
                  "column": 0,
                  "line": 4,
                },
              },
              "name": "border",
              "tags": Object {
                "unsupportedShorthand": true,
                "validSlot": null,
              },
              "type": "Property",
              "value": "1px solid red",
            },
          ],
          "scopes": Array [
            Object {
              "isLocal": false,
              "isSystem": false,
              "name": "when",
              "properties": Array [
                Object {
                  "loc": Object {
                    "end": Object {
                      "column": 3,
                      "line": 5,
                    },
                    "start": Object {
                      "column": 0,
                      "line": 5,
                    },
                  },
                  "name": "when",
                  "tags": Object {
                    "scope": "",
                    "shouldBeSlot": true,
                    "validSlot": null,
                  },
                  "type": "Property",
                  "value": "",
                },
                Object {
                  "loc": Object {
                    "end": Object {
                      "column": 8,
                      "line": 6,
                    },
                    "start": Object {
                      "column": 0,
                      "line": 6,
                    },
                  },
                  "name": "color",
                  "tags": Object {
                    "style": true,
                    "validSlot": null,
                  },
                  "type": "Property",
                  "value": "red",
                },
              ],
              "slotIsNot": undefined,
              "slotName": undefined,
              "value": "props.when",
            },
            Object {
              "isLocal": false,
              "isSystem": false,
              "name": "when",
              "properties": Array [
                Object {
                  "defaultValue": "stuff",
                  "loc": Object {
                    "end": Object {
                      "column": 11,
                      "line": 7,
                    },
                    "start": Object {
                      "column": 0,
                      "line": 7,
                    },
                  },
                  "name": "when",
                  "slotName": "",
                  "tags": Object {
                    "scope": "stuff",
                    "shouldBeSlot": true,
                    "slot": true,
                    "validSlot": true,
                  },
                  "type": "Property",
                  "value": "props.when",
                },
                Object {
                  "loc": Object {
                    "end": Object {
                      "column": 21,
                      "line": 8,
                    },
                    "start": Object {
                      "column": 0,
                      "line": 8,
                    },
                  },
                  "name": "backgroundColor",
                  "tags": Object {
                    "style": true,
                    "validSlot": null,
                  },
                  "type": "Property",
                  "value": "purple",
                },
                Object {
                  "loc": Object {
                    "end": Object {
                      "column": 10,
                      "line": 9,
                    },
                    "start": Object {
                      "column": 0,
                      "line": 9,
                    },
                  },
                  "name": "color",
                  "tags": Object {
                    "style": true,
                    "validSlot": null,
                  },
                  "type": "Property",
                  "value": "green",
                },
              ],
              "slotIsNot": false,
              "slotName": "",
              "value": "props.when",
            },
          ],
          "type": "Block",
        },
      ],
      "is": "Warning",
      "isBasic": true,
      "isGroup": true,
      "isList": false,
      "loc": Object {
        "end": Object {
          "column": 10,
          "line": 9,
        },
        "start": Object {
          "column": 0,
          "line": 0,
        },
      },
      "name": "Vertical",
      "properties": Array [],
      "scopes": Array [],
      "type": "Block",
    },
  ],
  "warnings": Array [
    Object {
      "line": "marginTop <",
      "loc": Object {
        "end": Object {
          "column": 10,
          "line": 3,
        },
        "start": Object {
          "column": 0,
          "line": 3,
        },
      },
      "type": "Add a default value to \\"marginTop\\" like: \\"marginTop < default value\\"",
    },
    Object {
      "line": "border 1px solid red",
      "loc": Object {
        "end": Object {
          "column": 19,
          "line": 4,
        },
        "start": Object {
          "column": 0,
          "line": 4,
        },
      },
      "type": "The shorthand border isn't supported. You need to expand it like:
borderWidth 1
borderStyle solid
borderColor red",
    },
    Object {
      "line": "when",
      "loc": Object {
        "end": Object {
          "column": 3,
          "line": 5,
        },
        "start": Object {
          "column": 0,
          "line": 5,
        },
      },
      "type": "This when has no condition assigned to it. Add one like: \\"when <isCondition\\"",
    },
    Object {
      "line": "backgroundColor purple",
      "loc": Object {
        "end": Object {
          "column": 21,
          "line": 8,
        },
        "start": Object {
          "column": 0,
          "line": 8,
        },
      },
      "type": "You're missing a base prop for backgroundColor. Add it before all whens on the block.",
    },
  ],
}
`;
